tensorflow/compiler/xla/service/algebraic_simplifier.cc

```diff
@@ -1665,6 +1668,265 @@ StatusOr<HloInstruction*> AlgebraicSimplifierVisitor::OptimizeDotOfGather(
   return memoized_lookup;
 }
 
+// This function tries to transform
+//   dot(reshape(transpose(A)), Const) to
+//   dot(reshape(A), reshape(transpose(reshape(Const)))),
+// so that the reshape and transpose on the Const side can be constant folded.
+//
+// The basic idea is that since the accumulation in the dot operation is
+// associative, so as long as we permute the elements of the contracting
+// dimensions on both sides of the dot in the same way, the result of the
+// dot is not affected.
+StatusOr<HloInstruction*>
+AlgebraicSimplifierVisitor::OptimizeDotOfReorderContractingDims(
+    HloInstruction* dot) {
+  HloInstruction* lhs_reshape;
+  HloInstruction* lhs_transpose;
+  HloInstruction* lhs_input;
+  HloInstruction* rhs_constant;
+
+  // This transformation assumes layout is not assigned yet.
+  if (options_.is_layout_sensitive()) {
+    return nullptr;
+  }
+
+  // Require single contracting dim to make the implementation easier to
+  // track contracting dims.
+  const auto& dnums = dot->dot_dimension_numbers();
+  if (dnums.lhs_contracting_dimensions_size() != 1) {
+    return nullptr;
+  }
+
+  // We allow constant on either side of dot. To simplify the checking and
+  // transformation, if lhs is constant we virtually swap lhs and rhs, so
+  // that lhs is always transpose and reshape. Before we create new dot,
+  // we will make sure the new lhs and rhs match the original ones.
+  if (!Match(dot, m::Dot().WithBinaryOperandsAnyOrder(
+                      m::Reshape(&lhs_reshape, m::Transpose(&lhs_transpose,
+                                                            m::Op(&lhs_input))),
+                      m::Constant(&rhs_constant)))) {
+    return nullptr;
+  }
+  int64 lhs_contracting_dim = dnums.lhs_contracting_dimensions(0);
+  int64 rhs_contracting_dim = dnums.rhs_contracting_dimensions(0);
+  if (dot->operand(0)->IsConstant()) {
+    std::swap(lhs_contracting_dim, rhs_contracting_dim);
```

Comment:
In the example I was suggesting that we actually modify the `dnums` variable, and that we say that at each step we have a valid dot.  I think that will be much more clear than having these variables like `lhs_reshape_contracting_dims` below.

I did not review past this point.

Reply:
Let me try to actually modify the dnums variable.